site/content/blog/2020-06-10-opentracing.md

```diff
@@ -0,0 +1,619 @@
+      return function(...args) {
+        console.log(`starting '${prop}'`);
+        const t1 = window.performance.now();
+
+        const res = target[prop](...arguments);
+        const t2 = window.performance.now();
+
+        console.log(`'${prop}' took ${t2 - t1}ns`);
+        return res;
+      };
+    }
+  }
+};
+
+const proxy = new Proxy(cow, tracerHandler);
+const proxy2 = new Proxy(proxy, timerHandler);
+console.log(proxy2.feed(20));
+console.log(proxy2.speak(3));
+```
+
+Unfortunately, in Go, there's no way to do this in a performant manner, and the regular approach would involve using reflection which can seriously impact performance on the underlying code.
+
+## Our solution
+
+The implementation of the decorator pattern we chose involved three parts:
+
+1. Struct embedding
+2. Code parsing using AST
+3. Code generation using templates
+
+### Struct embedding
+
+Quoting [Go FAQ](https://golang.org/doc/faq#Is_Go_an_object-oriented_language)
+> Although Go has types and methods and allows an object-oriented style of programming, there is no type hierarchy. The concept of "interface" in Go provides a different approach that we believe is easy to use and in some ways more general. There are also ways to embed types in other types to provide something analogous **but not identical** to subclassing
+
+```go
+type Animal struct{
+	Name string
+}
+
+type Cow struct{
+	Animal
+}
+
+func (c Cow) Speak() {
+	fmt.Printf("Moo, I am a %s", c.Animal.Name)
+}
+
+func main() {
+	a := Animal{Name:"Cow"}
+	c := Cow{Animal:a}
+	c.Speak()
+}
+```
+
+How does struct embedding help us in the implementation of a decorator pattern?
+
+```go
+
+type Speaker interface {
+	Speak(x int)
+}
+
+type Animal struct {
+	Name string
+}
+
+type TraceAnimal struct {
+	Speaker
+}
+
+type MeasureAnimal struct {
+	Speaker
+}
+
+func (c Animal ) Speak(x int) {
+	fmt.Println(strings.Repeat("I am a " + c.Name + " ",x))
+}
+
+func (c TraceAnimal) Speak(x int) {
+	fmt.Printf("Running Speak(x) function with x=%d!\n",x)
+	c.Speaker.Speak(x)
+}
+
+func (c MeasureAnimal) Speak(x int) {
+	fmt.Println("Timing Speak() function...")
+	t := time.Now()
+	c.Speaker.Speak(x)
+	
+	fmt.Printf("Speak(%d) took %s\n", x, time.Since(t))
+}
+
+func main() {
+	a := Animal{Name: "Cow"}
+	c := TraceAnimal {Speaker: a}
+	d := MeasureAnimal{Speaker: c}
+	d.Speak(2)
+}
+```
+
+Running the following code will yield:
+
+```
+Timing Speak() function...
+Running Speak(x) function with x=2!
+I am a Cow I am a Cow 
+Speak(2) took 0s
+```
+
+So we've basically implemented two decorators over original `Speak()` method. First we started timing the execution in `MeasureAnimal`, then passed it to `TraceAnimal`, which in turn called the actual `Speak()` implementation.
+
+This works great and stays performant since we don't use any dynamic techniques here like reflection, however this is very verbose and requires us to write a lot of wrapper code, and that's no fun at all. We can do better!
```

Comment:
```suggestion
This works great and stays performant since we don't use any dynamic techniques here like reflection, however this is very verbose and requires us to write a lot of wrapper code, and that's no fun at all. We can do better!
```
```suggestion
This works great and stays performant since we don't use any dynamic techniques such as reflection. However this is very verbose and requires us to write a lot of wrapper code - and that's no fun at all. 

We can do better!
```